velox/type/fbhive/HiveTypeParser.cpp

```diff
@@ -104,19 +104,20 @@ Result HiveTypeParser::parseType() {
           std::atoi(precision.value.data()), std::atoi(scale.value.data()))};
     } else if (nt.metadata->tokenString[0] == "date") {
       return Result{DATE()};
-    }
-    auto scalarType = createScalarType(nt.typeKind());
-    VELOX_CHECK_NOT_NULL(
-        scalarType, "Returned a null scalar type for ", nt.typeKind());
-    if (nt.metadata->tokenType == TokenType::String &&
+    } else if (
+        nt.metadata->tokenType == TokenType::String &&
         lookAhead() == TokenType::LeftRoundBracket) {
       eatToken(TokenType::LeftRoundBracket);
       Token length = nextToken();
       VELOX_CHECK(
           isPositiveInteger(length.value.toString()),
           "Varchar length must be a positive integer");
       eatToken(TokenType::RightRoundBracket);
+      return Result{VARCHAR(std::atoi(length.value.data()))};
```

Comment:
Does Hive support varchar(n) type? If not, let's not add this logic.

Reply:
Yes, it does support VarcharN: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types#LanguageManualTypes-StringsstringStrings
It uses the `string` type as unlimited varchar and the `Varchar(1)` to `Varchar(65553)` as length limited varchar.

As a result the  parser understands both types. If the length token is not present it creates an length unbounded varchar.

It may not be strictly necessary to add this in this PR but I think it would lay the ground for it to work out of the box.

Reply:
@czentgr Thank you for clarifying. It would be clearer to make this change in a separate PR, add a test, and include the above explaination in the PR description.

Reply:
I'm revising my earlier comment. I believe this change is required to support VARCHARN from the beginning. When the HiveConnector is set up it compares the hive column type to the velox type used to hold the data. Both of these could come from the plan (I saw them being referenced there). And if we can't parse both, type and hive_type and generate the equivalent type interactions with the table (insert, read, etc) fail.

Example, a table with a column created as VARCHAR(20). Because the hive type is not parsed but the presto type (now parsing the length) is we get a mismatch.

```
Query 20240828_174714_00010_cnaav failed: dataType_->equivalent(*hiveType_) data type VARCHAR(20) and hive type VARCHAR do not match
```

The hive_type in this case was `VARCHAR(20)` and so we can make an exact equivalent check on type and length. But, of course, VARCHAR(20) would be compatible with (but not equivalent to unbounded VARCHAR too but we probably should try to go for a proper parse and matching.

From the presto plan for an INSERT:
```
:[{"@type":"hive-hadoop2","name":"col1","hiveType":"varchar(20)","typeSignature":"varchar(20)","hiveColumnIndex":0,"columnType":"REGULAR","requiredSubfields":[]}],"pageSinkMetadata":{"schemaTableName":{"schema":"default","table":"varchar_limit"},"table":{"databaseName":"default","tableName":"varchar_limit","owner":"czentgr","tableType":"MANAGED_TABLE","storage":{"storageFormat":{"serDe":"org.apache.hadoop.hive.ql.io.orc.OrcSerde","inputFormat":"org.apache.hadoop.hive.ql.io.orc.OrcInputFormat","outputFormat":"org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat"}
```

`"name":"col1","hiveType":"varchar(20)","typeSignature":"varchar(20)"`